docs/cudf/source/developer_guide/pylibcudf.md

```diff
@@ -96,6 +96,28 @@ There are a couple of notable points from the snippet above:
 - The object returned from libcudf is immediately converted to a pylibcudf type.
 - `cudf::gather` accepts a `cudf::out_of_bounds_policy` enum parameter. `OutOfBoundsPolicy` is an alias for this type in pylibcudf that matches our Python naming conventions (CapsCase instead of snake\_case).
 
+## Testing
+
+Some guidelines on what should be tested:
+- Tests SHOULD comprehensively cover the API, including all possible combinations of arguments required to ensure good test coverage.
```

Comment:
Should we include a word about testing over a range of types here? The rule I've been following is that if the type is an explicit argument, then all supported types should be tested. However many APIs are supported for a range of types even though the type isn't an argument - scatter works for ints, floats, etc. Do we need to test all of them? Is it okay to just test one?

Reply:
Yes absolutely, we should figure out how we want to treat these. I worry that trying to test the Cartesian product of all types will lead to a prohibitively large number of tests, but maybe that's OK as long as we keep the runtime per test minimal? I know you said you've got thousands in your binops PR and that's taking a couple of minutes. Have you tried following the guidelines on fixtures that I wrote up in this PR? If not, can you see if that makes much of a difference?

Reply:
One compromise I've gone with in the past is to test over a matrix of "kind" rather than the full matrix. For instance, `{"uint64", "int64",  "float64", "string", "bool", "List", "Struct"}`. Might save a few orders of magnitude. What do you think?

Reply:
Yeah, that seems like a reasonable compromise to me. That way we are probably guaranteed to hit every distinct code path.

Another way of thinking about it: what we discussed was that pylibcudf should be focused on testing every code path and ensuring that all the logic that isn't testing in libcudf gets tested. All of the internal bits of pylibcudf involved in things like converting columns from libcudf, for example, need to be tested for every type since there is definitely type-specific logic there around converting different types of buffers etc. However, I don't think we need to test every user-facing API for every type because most APIs don't differ at all for different types in pylibcudf, only in libcudf. IOW something like copying a column is identical for string and float columns in pylibcudf up to differences in libcudf.

Thinking this way does require us to be a more cognizant of possible divergences in pylibcudf though, and it opens up an increased possibility for future error if we create new code branches in the future. A possible compromise would be to pick a couple of modules that we know exercise the internals sufficiently and test them for all types, then for others only test a sufficient subset of types.

Reply:
I went with basically this compromise. The chosen types are now standardized via a session-scope fixture, which should make it much easier to conform to these decisions across the whole test suite.